update/sync/syncHeadersByHash.go

```diff
@@ -0,0 +1,194 @@
+package sync
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/update"
+)
+
+type missingHeadersByHash struct {
+	mutMissingHdrs sync.Mutex
+	mapHeaders     map[string]data.HeaderHandler
+	mapHashes      map[string]struct{}
+	pool           dataRetriever.HeadersPool
+	storage        update.HistoryStorer
+	chReceivedAll  chan bool
+	marshalizer    marshal.Marshalizer
+	stopSyncing    bool
+	epochToSync    uint32
+	syncedAll      bool
+	requestHandler process.RequestHandler
+}
+
+// ArgsNewMissingHeadersByHashSyncer defines the arguments needed for the sycner
+type ArgsNewMissingHeadersByHashSyncer struct {
+	Storage        storage.Storer
+	Cache          dataRetriever.HeadersPool
+	Marshalizer    marshal.Marshalizer
+	RequestHandler process.RequestHandler
+}
+
+// NewMissingheadersByHashSyncer creates a syncer for all missing headers
+func NewMissingheadersByHashSyncer(args ArgsNewMissingHeadersByHashSyncer) (*missingHeadersByHash, error) {
+	if check.IfNil(args.Storage) {
+		return nil, dataRetriever.ErrNilHeadersStorage
+	}
+	if check.IfNil(args.Cache) {
+		return nil, update.ErrNilCacher
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, dataRetriever.ErrNilMarshalizer
+	}
+	if check.IfNil(args.RequestHandler) {
+		return nil, process.ErrNilRequestHandler
+	}
+
+	p := &missingHeadersByHash{
+		mutMissingHdrs: sync.Mutex{},
+		mapHeaders:     make(map[string]data.HeaderHandler),
+		mapHashes:      make(map[string]struct{}),
+		pool:           args.Cache,
+		storage:        args.Storage,
+		chReceivedAll:  make(chan bool),
+		requestHandler: args.RequestHandler,
+		stopSyncing:    true,
+		syncedAll:      false,
+		marshalizer:    args.Marshalizer,
+	}
+
+	p.pool.RegisterHandler(p.receivedHeader)
+
+	return p, nil
+}
+
+// SyncMissingHeadersByHash syncs the missing headers
+func (m *missingHeadersByHash) SyncMissingHeadersByHash(
+	shardIDs []uint32,
+	headersHashes [][]byte,
+	waitTime time.Duration,
+) error {
+	_ = process.EmptyChannel(m.chReceivedAll)
+
+	requestedMBs := 0
+	m.mutMissingHdrs.Lock()
+	m.stopSyncing = false
+	for index, hash := range headersHashes {
+		m.mapHashes[string(hash)] = struct{}{}
+		header, ok := m.getHeaderFromPoolOrStorage(hash)
+		if ok {
+			m.mapHeaders[string(hash)] = header
+			continue
+		}
+
+		requestedMBs++
+		m.requestHandler.RequestShardHeader(shardIDs[index], hash)
+	}
+	m.mutMissingHdrs.Unlock()
+
+	var err error
+	defer func() {
+		m.mutMissingHdrs.Lock()
+		m.stopSyncing = true
```

Comment:
Critical section in defer is a bit uncommon I think. Can you use `core/atomic.Flag` for `stopSyncing`?

Reply:
changed